### ISSUE TITLE
Infinite loop in 2.0 when value type has static method Value.of(List<?>)

### ISSUE THREAD

**Issue by bgertzfield** [immutables/immutables#102]
This value type works in Immutables.org 1.0 but creates an infinite loop in 2.0:

```
% cat Test.java
import org.immutables.value.Value;

import java.util.ArrayList;
import java.util.List;

@Value.Immutable
public abstract class Test {
  @Value.Parameter
  public abstract List<String> foo();

  public static Test of(List<String> foo) {
    return ImmutableTest.of(foo);
  }

  public static void main(String[] args) {
    System.out.format("%s\n", Test.of(new ArrayList<>()));
  }
}
% javac -processorpath value-2.0.6.jar -cp value-2.0.6.jar Test.java
% java -cp value-2.0.6.jar:. Test                                   
Exception in thread "main" java.lang.StackOverflowError
        at Test.of(Test.java:12)
        at Test.of(Test.java:12)
        at Test.of(Test.java:12)
        at Test.of(Test.java:12)
(snip)
```

This is because 1.x created a method `ImmutableTest.of(List<String> ...)` where 2.x creates a method `ImmutableTest.of(Iterable<String> ...)`:

```
  /**
   * Construct new immutable {@code Test} instance.
   * @param foo value for {@code foo}
   * @return immutable Test instance
   */
  public static ImmutableTest of(Iterable<String> foo) {
    return new ImmutableTest(foo);
  }
```

It looks like Java allows this but decides the call from the parent class to `ImmutableTest.of(foo)` binds better to the `List<?>` version than the `Iterable<String>` version.

**Comment by elucash** [comment]
Thank you for the report! Having iterable instead of list is beneficial as it reduces conversions being more liberal, I would be upset to roll this this feature back, moreover we have a code which depends on it. That aside, nothing else I can see could be fixed in Immutables.
On a practical sense here we are actually hitting java's peculiarities with how method resolution works. 

1. _To fix this occurrence_: change forwarding method
   + `Test of(List<String> foo)` to `Test of(Iterable<String> foo)`
   + Another alternative would be to have `create` method to forward to `of`
   + You can also alleviate the need for some of forwarding methods by using "reverse style" like `NewBuckStyle`

2. _To reveal other occurences_: enable warning/error for "The static method of(List<String>) from the type Test should be accessed directly" to prohibit access to static methods. I really think that java programmers should have this warning enabled in all cases. I'm not sure if Javac's _lint:static_ works for this, but ECJ and some other tools provide this check.

wdyt

**Comment by bgertzfield** [comment]
Yeah, it's definitely easy to work around this. I wanted to reproduce it in case you could think of some workaround Immutables could do (say, providing both `Iterable<T>` and `List<T>` interfaces or something).

It is pretty scary to run into a runtime-only stack overflow from upgrading a library, though..

**Comment by elucash** [comment]
The only thing that could possibly help on the side of Immutables is to generate overload. But I'm curious if this would be actually useful in a long run, and not be dead weight.

On one hand it will be redundant and largely unused, on the other hand we will need to generate only one overload and only if one or more `@Parameter`s are collection.

```java
static Z of(Iterable<? extends X> xs, int y) {
  ...// current implementation 
}

static Z of(List<X> xs, int y) {
  return of((Iterable<? extends X> x) xs, y);
}
```

any thoughts?

**Comment by trask** [comment]
At least the breakage is in a major version upgrade.  Ideally you'd keep `of(List<X>)` and mark that method `@Deprecated` for one release, but that won't work here b/c people will get spammed with deprecated warning messages because java method resolution will bind to `of(List)` instead of `of(Iterable)` if called with a `List` arg.  So I think drop it now or keep the strange looking overload forever...

**Comment by elucash** [comment]
Great points, Trask! So either create that overload or not: we will not be able to phase it out.

**Comment by elucash** [comment]
Here we need a sanity check. I've included logic to generate additional constructor, there should be not a lot of cases when it will be generated on average (constructor with collection attribute). We will not have any chance to remove it later. Can you, please, check the snapshots from sonatype oss repository (or locally built from master) to see and evaluate constructor?

**Comment by elucash** [comment]
Released in 2.0.7
